backend/pipelines/retriever.py

```diff
@@ -1,12 +1,104 @@
-from .pipeline_steps import PipelineSteps
+from backend.pipelines.pipeline_steps import PipelineSteps
+from backend.exceptions import CheckFailError, PipelineFailError
+from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast, AutoModel
+from elasticsearch import Elasticsearch, exceptions
 
 
 class Retriever(PipelineSteps):
-    def __init__(self, *args, **kwargs):
-        pass
+    def __init__(
+        self,
+        es: Elasticsearch,
+        elasticsearch_index: str,
+        contriever_model: AutoModel,
+        contriever_tokenizer: PreTrainedTokenizer | PreTrainedTokenizerFast,
+        size: int,
+        contriever_weight: int,
+    ):
+        self.es = es
+        self.elasticsearch_index = elasticsearch_index
+        self.contriever_model = contriever_model
+        self.contriever_tokenizer = contriever_tokenizer
+        self.size = size
+        self.contriever_weight = contriever_weight
 
-    def run(self):
-        pass
+    def run(self, data: dict) -> dict | PipelineFailError:
+        """Runs the retriever pipeline step.
 
-    def data_check(self):
-        pass
\ No newline at end of file
+        Args:
+            data (dict): The input data as a dictionary.
+
+        Raises:
+            PipelineFailError: If the pipeline step fails.
+
+        Returns:
+            dict: The modified data after executing the pipeline step.
+        """
+        try:
+            query_embedding = self.get_contriever_vector(data["query"]).tolist()[0]
+
+            body = {
+                "size": self.size,
+                "query": {
+                    "script_score": {
+                        "query": {"match": {"document": data["query"]}},
+                        "script": {
+                            "source": f"_score + {self.contriever_weight} * (cosineSimilarity(params.query_embedding, 'embedding') + 1.0)",
+                            "params": {"query_embedding": query_embedding},
+                        },
+                    }
+                },
+            }
+
+            s = self.es.search(index=self.elasticsearch_index, body=body)
+            contexts = list(s["hits"]["hits"])
+            data["official_contexts"] = [
+                context["_source"]["official_document"] for context in contexts
+            ]
+            data["lemmatized_contexts"] = [
+                context["_source"]["document"] for context in contexts
+            ]
```

Comment:
Can you add the addtional headers as well? Such as `_score`, `file_name`, `h1`, `h2` and `h3`. And make it in one `for`, because in this way we iterate through all context for each variable, instead of once.